Update Azure.Provisioning to latest

eerhardt opened the pull request:
## Description

Respond to latest renames getting the API ready for GA release.

Fix https://github.com/dotnet/aspire/issues/6376

I'll use this PR until we get stable versions from @tg-msft.

Also fix an issue when using dashes `-` in resource names and using `AspireV8ResourceNamePropertyResolver`. We are prematurely converting the dash to an underscore, and the Azure resource name is not the same as it was in v8. This needs to be fixed together because the APIs we need to fix the issue are being made public in this new Azure.Provisioning version.

Fix #6474

## Checklist

- Is this feature complete?
  - [x] No. Follow-up changes expected. Still need stable versions plus more changes.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to aspire-docs issue: https://github.com/dotnet/docs-aspire/issues/1652

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6390)

eerhardt committed:
Update Azure.Provisioning to latest

Respond to latest renames getting the API ready for GA release.

Fix #6376

eerhardt committed:
Rename properties to match new type names.

eerhardt commented on NuGet.config:
```diff
@@ -17,6 +17,7 @@
     <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
     <add key="dotnet-libraries" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries/nuget/v3/index.json" />
     <add key="dotnet9-transport" value="https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet9-transport/nuget/v3/index.json" />
+    <add key="azure-sdk-devfeed" value="https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json" />
```
These NuGet.config changes won't be merged. We won't merge this PR until the `1.0.0` versions are on nuget.org.

eerhardt commented on src/Aspire.Hosting.Azure/PublicAPI.Unshipped.txt:
```diff
@@ -17,18 +17,18 @@ Aspire.Hosting.Azure.AspireV8ResourceNamePropertyResolver
 Aspire.Hosting.Azure.AzureProvisioningOptions
 Aspire.Hosting.Azure.AzureProvisioningOptions.AzureProvisioningOptions() -> void
-Aspire.Hosting.Azure.AzureProvisioningOptions.ProvisioningContext.get -> Azure.Provisioning.ProvisioningContext!
+Aspire.Hosting.Azure.AzureProvisioningOptions.ProvisioningBuildOptions.get -> Azure.Provisioning.ProvisioningBuildOptions!
```
These API changes/renames are probably the most interesting changes in this PR.

eerhardt committed:
Add azure dev feed to test configs

eerhardt committed:
Use IsAccessKeyAuthenticationDisabled property

eerhardt committed:
Update to the latest dev build.

This hits an issue with `${null}` that is being fixed with https://github.com/Azure/azure-sdk-for-net/pull/46742.

eerhardt commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -886,8 +885,7 @@ private void AddContainerRegistryParameters(ContainerAppConfiguration app)
     {
-        var formatStringBuilder = new StringBuilder();
-        var arguments = new List<BicepValue<string>>();
+        var bicepStringBuilder = new BicepStringBuilder();
```
FYI - @davidfowl. We aren't able to get rid of our `Format` string parsing code from ReferenceExpression. But we are able to use the new `BicepStringBuilder` here.

eerhardt replied:
I think I could get rid of the BicepValueFormattableString class and just pass the `ReferenceExpression` around. No need in creating a new FormattableString object.

davidfowl replied:
We should be able to if you move this logic into the caller right?

eerhardt replied:
Yes, I can build up the BicepStringBuilder earlier and not have the BicepValueFormattableString at all.

davidfowl replied:
Ah yea, ok.

eerhardt commented:
Expecting 1 test failure due to https://github.com/Azure/azure-sdk-for-net/pull/46742

eerhardt committed:
Update to latest version

eerhardt committed:
Fix ContainerAppExtensions for new Azure.Provisioning version

eerhardt committed:
Simplify BicepValueFormattableString

eerhardt committed:
Simplify ReferenceExpression to BicepValue<string> logic

eerhardt committed:
Fix Azure Redis test for new property order

eerhardt committed:
Update to latest Azure.Provisioning version.

eerhardt commented:
This is now ready for review. We have all the changes from the Azure.Provisioning libraries we need for 9.0.0 GA.

eerhardt commented on src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs:
```diff
@@ -256,9 +256,9 @@ private static void CreateSqlServer(
         foreach (var databaseNames in databases)
         {
-            var identifierName = Infrastructure.NormalizeIdentifierName(databaseNames.Key);
+            var BicepIdentifier = Infrastructure.NormalizeBicepIdentifier(databaseNames.Key);
```
```suggestion
            var bicepIdentifier = Infrastructure.NormalizeBicepIdentifier(databaseNames.Key);
```

eerhardt commented on src/Aspire.Hosting.Azure.Sql/AzureSqlExtensions.cs:
```diff
@@ -256,9 +256,9 @@ private static void CreateSqlServer(
+            var BicepIdentifier = Infrastructure.NormalizeBicepIdentifier(databaseNames.Key);
             var databaseName = databaseNames.Value;
-            var sqlDatabase = new SqlDatabase(identifierName)
+            var sqlDatabase = new SqlDatabase(BicepIdentifier)
```
```suggestion
            var sqlDatabase = new SqlDatabase(bicepIdentifier)
```

eerhardt committed:
Apply suggestions from code review

eerhardt commented:
/backport to release/9.0

eerhardt committed:
Update playground bicep

eerhardt commented:
/backport to release/9.0

eerhardt committed:
Fix issue in AspireV8ResourceNamePropertyResolver when resource name has a dash.

In .NET Aspire 8.x, when the Aspire resource name has dashes, we used those dashes in the Azure resource name. But we are incorrectly converting them to underscores.

Fix #6474

eerhardt commented:
/backport to release/9.0

eerhardt committed:
Allow customizing Azure ContainerApp resources with ProvisioningBuildOptions.

Fix #6496

eerhardt commented:
/backport to release/9.0

joperezr committed:
Update packages to stable

eerhardt committed:
Revert test nuget.config changes.

eerhardt committed:
Fix up variable name.

eerhardt commented:
This has all the changes from [[release/9.0] Update Azure.Provisioning to latest (dotnet/aspire#6489)](https://github.com/dotnet/aspire/pull/6489) merged into it. Looking for an approval so we can merge it into `main`.

davidfowl commented:
/azp run

davidfowl commented:
/azp run
